qutebrowser/browser/pdfjs.py

```diff
@@ -0,0 +1,151 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""pdf.js integration for qutebrowser."""
+
+import os
+
+from PyQt5.QtCore import QUrl
+
+from qutebrowser.browser import webelem
+from qutebrowser.utils import utils
+
+
+class PDFJSNotFound(Exception):
+
+    """Raised when no pdf.js installation is found."""
+
+    pass
+
+
+def generate_pdfjs_page(url):
+    """Return the html content of a page that displays url with pdfjs.
+
+    Returns a string.
+
+    Args:
+        url: The url of the pdf as QUrl.
+    """
+    viewer = get_pdfjs_res('web/viewer.html').decode('utf-8')
+    script = _generate_pdfjs_script(url)
+    html_page = viewer.replace(
+        '</body>', '</body><script>{}</script>'.format(script)
+    )
+    return html_page
+
+
+def _generate_pdfjs_script(url):
+    """Generate the script that shows the pdf with pdf.js.
+
+    Args:
+        url: The url of the pdf page as QUrl.
+    """
+    return (
+        'PDFView.open("{url}");\n'
+    ).format(url=webelem.javascript_escape(url.toString(QUrl.FullyEncoded)))
+
+
+def fix_urls(asset):
+    """Take a html page and replace each relative URL wth an absolute.
+
+    This is specialized for pdf.js files and not a general purpose function.
+
+    Args:
+        asset: js file or html page as string.
+    """
+    new_urls = {
+        'viewer.css': 'qute://pdfjs/web/viewer.css',
+        'compatibility.js': 'qute://pdfjs/web/compatibility.js',
+        'locale/locale.properties':
+            'qute://pdfjs/web/locale/locale.properties',
+        'l10n.js': 'qute://pdfjs/web/l10n.js',
+        '../build/pdf.js': 'qute://pdfjs/build/pdf.js',
+        'debugger.js': 'qute://pdfjs/web/debugger.js',
+        'viewer.js': 'qute://pdfjs/web/viewer.js',
+        'compressed.tracemonkey-pldi-09.pdf': '',
+        './images/': 'qute://pdfjs/web/images/',
+        '../build/pdf.worker.js': 'qute://pdfjs/build/pdf.worker.js',
+        '../web/cmaps/': 'qute://pdfjs/web/cmaps/',
+    }
+    for original, new in new_urls.items():
+        asset = asset.replace(original, new)
+    return asset
+
+
+SYSTEM_PDFJS_PATHS = [
+    '/usr/share/pdf.js/',  # Debian pdf.js-common
+    '/usr/share/javascript/pdf/',  # Debian libjs-pdf
+    os.path.expanduser('~/.local/share/qutebrowser/pdfjs/'),  # fallback
+]
+
+
+def get_pdfjs_res(path):
+    """Get a pdf.js resource in binary format.
+
+    Args:
+        path: The path inside the pdfjs directory.
+    """
+    path = path.lstrip('/')
+    content = None
+
+    # First try a system wide installation
+    # System installations might strip off the 'build/' or 'web/' prefixes.
+    # qute expects them, so we need to adjust for it.
+    names_to_try = [path, path[path.index('/') + 1:]]
```

Comment:
Since `pdfjs.py` expects the archive structure of the pdf.js archive, each file is expected to have at least a `web/` or `build/` prefix - with exception of the `LICENSE` file, which isn't currently used though (though I could add it as `qute://pdfjs/LICENSE` if you want.)